POEM_081.md

```diff
@@ -0,0 +1,404 @@
+
+                # update options outputs dict with new output dict
+                self.options['outputs'].update(out_dict)
+
+            elif isinstance(out, str):
+                # check if variable already exists in self.options['outputs']
+                # i.e. no repeated output variable names
+                if out in self.options['outputs']:
+                    raise Exception(f'Variable {out} already exists. Rename variable'
+                                    ' or delete copy of variable.')
+
+                # make dict with given var name as key and meta data from model_outputs
+                out_dict = {out: meta for _,meta in model_outputs.items()
+                            if meta['prom_name'].endswith(out)}
+
+                # check if provided variable appears more than once in model
+                if len(out_dict) > 1:
+                    raise Exception(f'Ambiguous variable {out} in outputs. To'
+                                    ' specify which one is desired, use a tuple'
+                                    ' with the promoted name and variable name'
+                                    ' instead [e.g. (prom_name, var_name)].')
+
+                 # checks if provided variable doesn't exist in model
+                elif len(out_dict) == 0:
+                    raise Exception(f'Variable {out} does not exist in model.')
+
+                # update options outputs dict with new output dict
+                self.options['outputs'].update(out_dict)
+
+            else:
+                raise Exception(f'Type {type(out)} is invalid for output. Must be'
+                                ' string or tuple.')
+
+    def setup(self):
+        inputs = self.options['inputs']
+        outputs = self.options['outputs']
+
+        # instantiate input/output name list for use in compute and 
+        # compute partials
+        self._input_names = []
+        self._output_names = []
+
+        # remove the `prom_name` from the metadata and then store it for each
+        # input and output
+        for var, meta in inputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_input(var, **meta)
+            # add prom_name back
+            meta['prom_name'] = prom_name
+            self._input_names.append(var)
+
+        for var, meta in outputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_output(var, **meta)
+            # add prom_name back
+            meta['prom_name'] = prom_name
+            self._output_names.append(var)
+
+            for ip in self._input_names:
+                self.declare_partials(of=var, wrt=ip)
+    
+    def compute(self, inputs, outputs):
+        p = self._subprob
+
+        # switch subproblem to use complex IO if in complex step mode
+        if self.under_complex_step != self._prev_complex_step:
+            if self.under_complex_step:
+                p.setup(force_alloc_complex=True)
```

Comment:
Allocation of complex vectors due to `force_alloc_complex` should only ever be done once, i.e., if we ever have to do it, we should just leave it that way from that point on instead of constantly allocating and deallocating the problem vectors.  We will still need to call problem.set_complex_step_mode(True/False) when it changes, but we don't have to keep doing the vector reallocation.